testsuite/oauth/authorize.py

```diff
@@ -0,0 +1,236 @@
+        }
+        if 'extra_launch_params' in self.config:
+            params.update(self.config['extra_launch_params'])
+
+        return params
+
+    def open(self):
+        """ Connect to the selenium webdriver.
+        """
+        self.browser = self._browser()
+
+    def close(self):
+        """ Close the virtual browser.
+        """
+        self.browser.quit()
+        self.display.stop()
+
+    def __enter__(self):
+        self.open()
+        return self.browser
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _browser(self):
+        """ Initialize a Chrome webdriver.
+        """
+        self.display = Display(visible=0, size=(800, 600))
+        self.display.start()
+
+        options = Options()
+        options.add_argument("--no-sandbox")
+
+        return webdriver.Chrome(chrome_options=options)
+
+    def _execute_step(self, step):
+        try:
+            elem = self.browser.find_element_by_css_selector(step['element'])
+        except NoSuchElementException as err:
+            if step.get('optional'):
+                return
+            raise ElementNotFoundException(str(err), self.browser)
+
+        # Make sure the element exists before we continue
+        if not elem.is_displayed() and step.get('optional'):
+            return
+        elif not elem.is_displayed():
+            raise ElementNotFoundException('Element is hidden', self.browser)
+
+        # Apply the action to the matched element.
+        # Only one action can be applied per step.
+        if 'send_keys' in step:
+            elem.send_keys(step['send_keys'])
+        elif 'click' in step:
+            elem.click()
+        else:
+            raise NoStepCommandException(step, self.browser)
+
+    def _validate_state(self, query):
+        assert 'state' in query, 'Missing state parameter.'
+
+        assert len(query['state']) == 1, 'Too many state parameters.'
+
+        assert query['state'][0] == self._state, \
+            'Returned state parameter does not match sent state.'
+
+    def _validate_code(self, query):
+        assert 'code' in query, 'Missing code parameter.'
+
+        assert len(query['code']) == 1, 'Too many code parameters.'
+
+
+class CurrentUrlContains(object):
+    """ An expectation that the browser's current url contains a
+    case-sensitive substring.
+
+    @see: http://selenium-python.readthedocs.io/waits.html
+
+    Returns:
+        True when the URL matches, False otherwise
+    """
+    def __init__(self, substring):
+        self.substring = substring
+
+    def __call__(self, driver):
+        return self.substring in driver.current_url
+
+
+class AuthorizationException(Exception):
+    """ An Error occured during the authorization process.
+    """
+    def __init__(self, message, browser):
+        # Save a screenshot so we can see what's up
+        path = 'testsuite/static/screenshots/{0}.png'.format(uuid.uuid4())
+        browser.save_screenshot(path)
```

Comment:
Note that in current status, I can get errors on the server side like:

```
    self._execute_step(step)
  File "/usr/src/app/testsuite/oauth/authorize.py", line 161, in _execute_step
    raise ElementNotFoundException('Element is hidden', self.browser)
testsuite.oauth.authorize.ElementNotFoundException: ('Element is hidden', '/static/screenshots/777e8c08-1844-42da-bf27-49c12c3a95f3.png')
```

but unless they occur in a "When I ask for authorization with the following override" scenario, they don't bubble up to the UI (and the UI spins forever). Would be good to emit a "failed" message" when any uncaught exception occurs, so the UI can render it.